flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessor.java

```diff
@@ -84,8 +84,9 @@ public DataInputStatus processInput() throws Exception {
 
         lastReadInputIndex = readingInputIndex;
         DataInputStatus inputStatus = inputProcessors[readingInputIndex].processInput();
+        inputSelectionHandler.updateStatus(inputStatus, readingInputIndex);
         inputSelectionHandler.nextSelection();
-        return inputSelectionHandler.updateStatus(inputStatus, readingInputIndex);
+        return inputSelectionHandler.calculateOverallStatus(inputStatus);
```

Comment:
1. Why did you have to reverse `updateStatus()`/`nextSelection()` order?
2. I hope that JIT will be clever enough to inline and optimise this code for `MORE_AVAILABLE` case (I don't know that. Maybe 1. will prevent that).
3. Can not we keep the `MORE_AVAILABLE` hot path as it was before? I'm not feeling confident if that's an issue or not, but I intentionally made `updateStatus()` the way it was, to ensure `MORE_AVAILABLE` path is as simple as possible.

Reply:
In had to reverse the order, because in `updateStatus` I update statuses for `END_OF_DATA`/`END_OF_PARTITION` which I need for the calculation in `nextSelection`.

Reply:
In that case I have a feeling that those methods should be restructured, maybe squashed together? More or less you could have kept the previous structure of `updateStatus()` but move calling `nextSelection()` inside it?

Reply:
I moved nextSelection inside of updateStatus, ptal.